cpp/src/arrow/flight/types.cc

```diff
@@ -83,6 +84,31 @@ Status FlightInfo::GetSchema(std::shared_ptr<Schema>* out) const {
   return Status::OK();
 }
 
+Location::Location() {
+  uri_ = std::shared_ptr<arrow::internal::Uri>(new arrow::internal::Uri());
+}
+
+Status Location::Parse(const std::string& uri_string, Location& location) {
+  return location.uri_->Parse(uri_string);
+}
+
+Status Location::ForGrpcInsecure(const std::string& host, const int port,
+                                 Location& location) {
+  std::stringstream uri_string;
+  uri_string << "grpc://" << host << ':' << port;
```

Comment:
Perhaps it can be fixed later, but this won't work for IPv6 numeric addresses, e.g. you need `grpc://[::1]:80` and not `grpc://::1:80`.

Reply:
uriparser (understandably) doesn't deal with URI construction, so we'd need recognize IPv6 addresses, or create a separate method for such addresses. Or perhaps just require that the user pass `[::1]`?

Reply:
Yes, we can require that for now.
(uriparser seems able to deal with URI construction, but the API looks a bit terrible)